### PR TITLE
fix empty byte array comparison

### PR DESCRIPTION
**Pull request by tserg**
### What I did

Fix for #2704.

### How I did it

Handle comparison of byte arrays (bytes, strings) that are `empty(String[N])` or `empty(Bytes[N])` by adding a flag to check for instances where both byte arrays are of equal max length and at least one value is `~empty`. If yes, check the length of the non-empty side against 0.

[UPDATE]

Comparison of byte arrays that are empty (i.e. "", b"") are updated to use the same codepath, instead of applying keccak to both values.

### How to verify it

See tests.

### Commit message

```
This commit fixes an issue with equality checks involving empty(byte arrays)
(i.e. empty(Bytes[N]), empty(String[N])). 

A new flag is introduced to check if comparison involves an empty(byte array).
If yes, check the length of the non-empty side against 0.

Additionally, another flag is introduced to check if comparison involves a byte array
that is empty (i.e. "", b""). If yes, this will take the same codepath as above. This 
eliminates the need to keccak both values, reducing the size of the contract.

Resolves: #2704
```

### Description for the changelog

Fix empty byte array comparison.t

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: vyper/codegen/expr.py [hunk 1]
```diff
@@ -890,10 +890,13 @@ def parse_Compare(self):
...
 
             length_mismatch = left.typ.maxlen != right.typ.maxlen
+            empty_eq_length_array = left.typ.maxlen == right.typ.maxlen and (
+                right.value == "~empty" or left.value == "~empty"
+            )
             left_over_32 = left.typ.maxlen > 32
             right_over_32 = right.typ.maxlen > 32
 
-            if length_mismatch or left_over_32 or right_over_32:
+            if length_mismatch or empty_eq_length_array or left_over_32 or right_over_32:
```
Comment (charles-cooper) [inline vyper/codegen/expr.py:899]: no need to use keccak if one side is `empty`, can just check that the length of the other side is 0.

**Review by charles-cooper — CHANGES_REQUESTED** [review]
nice, could you also add a couple tests for empty string literals (e.g. `b""`)? I think they follow the same codepath but not positive. also a test for Bytes where `N>32` could be good for completeness. (maybe parametrize to keep the amount of test code finite)

**Comment by charles-cooper** [comment]
@tserg the codegen is fixed in https://github.com/vyperlang/vyper/commit/2c73f8352635c0a433423a5b94740de1a118e508 but I would like to get the tests from this merged into master. Could you please rebase off master and just keep the added tests?

**Review by charles-cooper — APPROVED** [review]
